internal/servicename.go

```diff
@@ -21,27 +21,71 @@
 package internal
 
 import (
+	"errors"
 	"fmt"
 	"regexp"
-)
-
-var (
-	// Must starts with [a-z].
-	// Can contain [a-z] and 0-9] with non-consecutive dashes.
-	validNameRegex = regexp.MustCompile("^[a-z]+([a-z0-9]|[^-]-)*[^-]$")
 
-	// We disallow UUIDs explicitly (they would be accepted by validNameRegex alone)
-	uuidRegex = regexp.MustCompile("[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}")
+	"go.uber.org/multierr"
 )
 
-// ValidateServiceName returns an error if the given name is not a valid
-// service name.
+// We disallow UUIDs explicitly, though they're otherwise valid patterns.
+var _uuidRegexp = regexp.MustCompile("[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}")
+
+// ValidateServiceName returns an error if the given servive name is invalid.
+// Valid names are at least two characters long, start with [a-z], contain only
+// [0-9a-z] and non-consecutive hyphens, and end in [0-9a-z]. Furthermore,
+// names may not contain UUIDs.
 func ValidateServiceName(name string) error {
-	if !validNameRegex.MatchString(name) {
-		return fmt.Errorf("service name must begin with a letter and consist only of dash-delimited lower-case ASCII alphanumeric words")
+	if len(name) < 2 {
+		// Short names aren't safe to check any further.
+		return errors.New("service name must be at least two characters long")
```

Comment:
Really? All the subsequent validation uses multierr; this validation must pass, though, since the indexing logic in the other checks isn't safe otherwise. I *could* replicate this check in every other check as a short-circuit, but that doesn't seem like a step forward.